### ISSUE TITLE
80% times remote participants can't join room or share video and audio 

### ISSUE THREAD

**Issue by akourm910e** [twilio/twilio-video.js#917]
- [x] I have verified that the issue occurs with the latest twilio-video.js release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-video.js/blob/master/CHANGELOG.md).
 - [x] I reviewed the [Common Issues](https://github.com/twilio/twilio-video.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
 - [x] I verified that the [Quickstart](https://github.com/twilio/video-quickstart-js) application works in my environment.
 - [x] I am not sharing any [Personally Identifiable Information (PII)](https://www.twilio.com/docs/glossary/what-is-personally-identifiable-information-pii)
       or sensitive account information (API keys, credentials, etc.) when reporting this issue.

**Code to reproduce the issue:**

<template>
    <div class="teleconsultation-container">
        <div class="title-page">Téléconsultation</div>
        <Spinner color="#ec6283" v-show="loading"></Spinner>
        <div v-show="!joinVisio && !loading">
            <div class="row" style="display: flex">
                <div class="icon-container">
                    <i class="icons i-visio" />
                </div>
                <div class="explain-container">
                    <div class="hi">
                        Bonjour, vous êtes dans la salle d'attente de votre
                        téléconsultation, vous serez notifié de la connexion de
                        votre
                        {{
                            $userprofile.patientaccount ? "médecin" : "patient"
                        }}...
                    </div>
                    <div
                        v-if="activeRoom.localParticipant"
                        type="who-connected"
                    >
                        {{ activeRoom.localParticipant.identity }} est
                        <span class="green">connecté</span>
                    </div>
                    <div>
                        {{ participantName }} est
                        <span class="green" v-if="participants.length > 0"
                            >connecté</span
                        >
                        <span class="red" v-else>non connecté</span>
                    </div>
                    <div class="error-camera">
                        <el-tag type="danger" v-if="errorCamera"
                            >Il y a un problème avec votre caméra/microphone,
                            veuillez vérifier vos paramètres</el-tag
                        >
                    </div>

                    <div
                        class="participant-connected"
                        v-if="participants.length > 0"
                    >
                        <el-tag type="success">
                            Le
                            {{
                                $userprofile.patientaccount
                                    ? "médecin"
                                    : "patient"
                            }}
                            est connecté, cliquez sur le bouton ci-dessous pour
                            rejoindre la visioconférence
                        </el-tag>
                    </div>
                    <div class="test-camera-title" v-if="testCamera">
                        Nous testons votre caméra et votre microphone, patientez
                    </div>
                    <el-card
                        style="background: none"
                        v-loading="testCamera"
                        class="visioContainer"
                        v-if="testCamera"
                    >
                        <div id="previewTrack"></div>
                    </el-card>
                    <div class="actions-visio">

                        <el-button
                            v-if="joinVisio === false && !testCamera"
                            @click="activeVisio(activeRoom)"
                            :disabled="participants.length > 0 ? false : true"
                            type="danger"
                        >
                            Rejoindre la téléconsultation avec votre
                            {{
                                $userprofile.patientaccount
                                    ? "médecin"
                                    : "patient"
                            }}
                        </el-button>
                    </div>
                    <div class="help-visio">
                        Un problème avec votre visioconférence ?
                        <el-link @click="refresh" type="danger"
                            >Recharger la page</el-link
                        >
                    </div>
                </div>
            </div>
        </div>

        <div id="visioPart" v-show="joinVisio && !loading">
            <div>
                <div class="visioContainer">
                    <div id="localTrack"></div>

                    <div id="remoteTrack"></div>
                </div>
            </div>
        </div>
        <span v-if="joinVisio && !loading" slot="footer" class="dialog-footer">
            <el-button class="footer-right" type="danger" @click="endVisio()"
                >Terminer la téléconsultation</el-button
            >
        </span>
  
    </div>
</template>

<script>
import Spinner from "../spinner";
import { mapState, mapActions, mapGetters } from "vuex";
import { EventBus } from "../Event";
import Router from "vue-router";
import Twilio, {
    connect,
    createLocalTracks,
    createLocalVideoTrack
} from "twilio-video";
export default {
    data() {
        return {
            fullScreen: false,
            joinVisio: false,
            modalName: "VisioConference",
            loading: false,
            data: {},
            localTrack: false,
            remoteTrack: "",
            remoteMediaContainer: "",
            previewContainer: "",
            activeRoom: "",
            previewTracks: "",
            identity: "",
            roomName: "",
            username: "",
            logs: [],
            logCount: 0,
            participants: [],
            connectOptions: {},
            testCamera: false,
            errorTestCamera: "",
            errorCamera: false
        };
    },
    props: ["popupVisio", "username", "roomname", "participantName"],
    methods: {
        ...mapActions("actions", ["getTwilioToken"]),

        cancel() {
            this.hide();
        },
        validate() {
            this.$emit("onvalidate", this.$data);
        },
        show() {
            this.$modal.show(this.modalName);
        },
        hideModal() {
            this.$emit("hideModalVisio");
        },
        beforeOpen() {
            this.$emit("beforeopen", this.$data);
        },
        opened() {
            this.$emit("opened", this.$data);
        },
        beforeClose() {
            this.$emit("beforeclose", this.$data);
        },
        closed() {
            // this.popupVisio = false;
            this.$emit("closedVisioModal");
        },

        attachTrack(track, container) {
            container.appendChild(track.attach());
        },

        // Attach array of Tracks to the DOM.
        attachTracks(tracks, container) {
            const that = this;
            tracks.forEach(function(track) {
                that.attachTrack(track, container);
            });
        },
        // Detach given track from the DOM.
        detachTrack(track) {
            track.detach().forEach(function(element) {
                element.remove();
            });
        },

        previewCamera() {
            this.testCamera = true;
            const that = this;
            createLocalVideoTrack()
                .then(track => {
                    let localMediaContainer = document.getElementById(
                        "previewTrack"
                    );
                    localMediaContainer.appendChild(track.attach());
                    setTimeout(function() {
                        if (localMediaContainer) {
                            
                                track.stop();
                    
                            localMediaContainer = null;
                        }

                        that.$toasted.success(
                            "Votre caméra et microphone ont bien été détecté."
                        );
                        that.testCamera = false;
                    }, 3000);
                })
                .catch(error => {
                    setTimeout(function() {
                        console.error("Unable to access local media", error);
                        that.testCamera = false;
                        that.errorCamera = true;
                        that.$toasted.error(
                            "Impossible d'accéder à votre caméra et/ou votre microphone, veuillez vérifier vos paramètres"
                        );
                    }, 3000);
                });
        },

        refresh() {
            location.reload();
        },

        joinRoom(room_name) {
            console.log("ROOM NAME", room_name);
            // this.joinVisio = true;
            // this.fullScreen = true;
            this.loading = true;
            const that = this;
            this.getTwilioToken({
                username: this.username,
                roomname: this.roomName
            }).then(data => {
                that.roomName = null;
                console.log("TOKEN", data.data.response);
                const token = data.data.response.toString();
                this.connectOptions = {
                    name: room_name,
                    logLevel: "warn",
                    audio: true,
                    video: { width: 720 },
                    _useTwilioConnection: true
                };
                Twilio.connect(token, this.connectOptions).then(room => {
                    that.activeRoom = room;
                    that.$toasted.success(
                        "Vous avez rejoint la salle d'attente virtuelle"
                    );
                    that.activeRoom.participants.forEach(participant => {
                        console.log("PARTICIPANTS", participant.identity);
                        that.participants.push(participant.identity);
                        that.$toasted.success(
                            participant.identity + " est connecté"
                        );
                        participant.tracks.forEach(publication => {
                            if (publication.isSubscribed) {
                                const track = publication.track;
                                document
                                    .getElementById("remoteTrack")
                                    .appendChild(track.attach());
                            }
                        });
                    });

                    that.activeRoom.on("participantConnected", function(
                        participant
                    ) {
                        that.participants.push(participant.identity);
                        that.$toasted.success(
                            participant.identity +
                                " a rejoint la téléconsultation."
                        );
                        participant.on("trackSubscribed", track => {
                            document
                                .getElementById("remoteTrack")
                                .appendChild(track.attach());
                        });
                    });
                    that.activeRoom.on("participantDisconnected", function(
                        participant
                    ) {
                        that.participants = [];
                        that.$toasted.success(
                            participant.identity +
                                " a quitté la téléconsultation"
                        );
                        that.joinVisio = false;
                    });

                    that.activeRoom.on("disconnected", room => {
                        // Detach the local media elements
                        room.localParticipant.tracks.forEach(publication => {
                            const attachedElements = publication.track.detach();
                            attachedElements.forEach(element =>
                                element.remove()
                            );
                        });
                    });
                    that.activeRoom.localParticipant.videoTracks.forEach(
                        track => {
                            track.track.disable();
                        }
                    );
                    that.activeRoom.localParticipant.audioTracks.forEach(
                        track => {
                            track.track.disable();
                        }
                    );

                    console.log("ROOM ACTIVE", room);
                    that.roomName = room_name;
                    that.loading = false;
                });
            });
            this.loading = false;
        },

        // Successfully connected!
        activeVisio(room) {
            const that = this;
            this.joinVisio = true;

            var localTracksPromise = that.previewTracks
                ? Promise.resolve(that.previewTracks)
                : createLocalTracks({
                      audio: true,
                      video: { width: 640 }
                  });

            localTracksPromise.then(
                function(tracks) {
                    that.previewTracks = tracks;
                    that.previewContainer = document.getElementById(
                        "localTrack"
                    );
                    if (!that.previewContainer.querySelector("video")) {
                        that.attachTracks(tracks, that.previewContainer);
                    }
                },
                function(error) {
                    console.error("Unable to access local media", error);
                    that.$toasted.error(
                        "Impossible d'accéder à votre caméra et/ou votre microphone"
                    );
                }
            );

            room.participants.forEach(participant => {
                console.log("PARTICIPANTS", participant.identity);
                that.participants.push(participant.identity);
                that.$toasted.success(participant.identity + " est connecté");
                participant.tracks.forEach(publication => {
                    if (publication.isSubscribed) {
                        const track = publication.track;
                        document
                            .getElementById("remoteTrack")
                            .appendChild(track.attach());
                    }
                });
            });
            room.localParticipant.videoTracks.forEach(track => {
                track.track.enable();
            });

            room.localParticipant.audioTracks.forEach(track => {
                track.track.enable();
            });
            // When a Participant joins the Room, log the event.
        },

        // Detach the Participant's Tracks from the DOM.
        detachParticipantTracks(participant) {
            const that = this;
            var tracks = that.getTracks(participant);
            tracks.forEach(that.detachTrack);
        },

        getTracks(participant) {
            return Array.from(participant.tracks.values())
                .filter(function(publication) {
                    return publication.track;
                })
                .map(function(publication) {
                    return publication.track;
                });
        },

        disconnect() {
            this.activeRoom.disconnect();
        },

        endVisio() {
            const that = this;
            this.$confirm(
                "Voulez-vous terminer la téléconsultation ?",
                "Confirmation",
                {
                    confirmButtonText: "Confirmer",
                    cancelButtonText: "Annuler",
                    type: "warning"
                }
            )
                .then(() => {
                    that.$emit("closedVisioModal");
                    that.disconnect();
                })
                .catch(() => {});
        }
    },

    created() {
        const that = this;
        this.previewCamera();
        this.joinRoom(this.roomname);
        window.addEventListener("beforeunload", () => {
            that.disconnect();
        });
    },

    beforeDestroy() {
        this.disconnect();
    },

    component: {
        Spinner
    }
};
</script>

**Expected behavior:** 

LocalParticipant and remoteParticipants joins the same room, and can share video and audio

**Actual behavior:**

Everything going well when we try with my teammates. Audio and video works. 

In production, 80% times remote participants can't connect to the room. Or the remote participant can't share video and audio. 

**Software versions:**

 - [ Google Chrome 80 ] Browser(s):
 - [ Linux - xubuntu 19.04 ] Operating System:
 - [ 2.3 ] twilio-video.js:
 - [ Vue-cli 3 ] Third-party libraries (e.g., Angular, React, etc.):
